// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.
//
// The following only applies to changes made to this file as part of YugaByte development.
//
// Portions Copyright (c) YugaByte, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
// in compliance with the License.  You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software distributed under the License
// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
// or implied.  See the License for the specific language governing permissions and limitations
// under the License.
//
syntax = "proto2";

package yb.tserver;

option java_package = "org.yb.tserver";

import "yb/common/common_types.proto";
import "yb/common/transaction.proto";
import "yb/tablet/tablet_types.proto";
import "yb/tablet/operations.proto";
import "yb/tserver/tserver.proto";
import "yb/tserver/tserver_types.proto";

service TabletServerService {
  rpc Write(WriteRequestPB) returns (WriteResponsePB);
  rpc Read(ReadRequestPB) returns (ReadResponsePB);
  rpc VerifyTableRowRange(VerifyTableRowRangeRequestPB)
      returns (VerifyTableRowRangeResponsePB);

  rpc NoOp(NoOpRequestPB) returns (NoOpResponsePB);
  rpc ListTablets(ListTabletsRequestPB) returns (ListTabletsResponsePB);
  rpc GetLogLocation(GetLogLocationRequestPB) returns (GetLogLocationResponsePB);

  // Run full-scan data checksum on a tablet to verify data integrity.
  //
  // TODO: Consider refactoring this as a scan that runs a checksum aggregation
  // function.
  rpc Checksum(ChecksumRequestPB) returns (ChecksumResponsePB);

  rpc ListTabletsForTabletServer(ListTabletsForTabletServerRequestPB)
      returns (ListTabletsForTabletServerResponsePB);

  rpc ImportData(ImportDataRequestPB) returns (ImportDataResponsePB);
  rpc UpdateTransaction(UpdateTransactionRequestPB) returns (UpdateTransactionResponsePB);
  // Returns transaction status at coordinator, i.e. PENDING, ABORTED, COMMITTED etc.
  rpc GetTransactionStatus(GetTransactionStatusRequestPB) returns (GetTransactionStatusResponsePB);
  // Returns transaction status at participant, i.e. number of replicated batches or whether it was
  // aborted.
  rpc GetTransactionStatusAtParticipant(GetTransactionStatusAtParticipantRequestPB)
      returns (GetTransactionStatusAtParticipantResponsePB);
  rpc AbortTransaction(AbortTransactionRequestPB) returns (AbortTransactionResponsePB);
  // Change the status tablet id used for a transaction.
  rpc UpdateTransactionStatusLocation(UpdateTransactionStatusLocationRequestPB)
      returns (UpdateTransactionStatusLocationResponsePB);
  rpc UpdateTransactionWaitingForStatus(UpdateTransactionWaitingForStatusRequestPB)
      returns (UpdateTransactionWaitingForStatusResponsePB);
  rpc ProbeTransactionDeadlock(ProbeTransactionDeadlockRequestPB)
      returns (ProbeTransactionDeadlockResponsePB);
  rpc Truncate(TruncateRequestPB) returns (TruncateResponsePB);
  rpc GetTabletStatus(GetTabletStatusRequestPB) returns (GetTabletStatusResponsePB);
  rpc GetMasterAddresses(GetMasterAddressesRequestPB) returns (GetMasterAddressesResponsePB);

  rpc Publish(PublishRequestPB) returns (PublishResponsePB);

  rpc IsTabletServerReady(IsTabletServerReadyRequestPB) returns (IsTabletServerReadyResponsePB);

  rpc GetSplitKey(GetSplitKeyRequestPB) returns (GetSplitKeyResponsePB);

  rpc GetSharedData(GetSharedDataRequestPB) returns (GetSharedDataResponsePB);

  rpc GetTserverCatalogVersionInfo(GetTserverCatalogVersionInfoRequestPB)
      returns (GetTserverCatalogVersionInfoResponsePB);
}

message GetLogLocationRequestPB {
}

message GetLogLocationResponsePB {
  optional string log_location = 1;
}

message ChecksumRequestPB {
  reserved 1, 2, 3, 4, 5;

  optional bytes tablet_id = 6;
  optional YBConsistencyLevel consistency_level = 7;
}

message ChecksumResponsePB {
  reserved 3, 4, 5;

  // Error message, if any.
  optional TabletServerErrorPB error = 1;

  // The (possibly partial) checksum of the tablet data.
  // This checksum is only complete if 'has_more_results' is false.
  optional uint64 checksum = 2;
}

message ListTabletsForTabletServerRequestPB {
}

message ListTabletsForTabletServerResponsePB {
  // An entry for each tablet.
  message Entry {
    optional string table_name = 1;
    optional bytes tablet_id = 2;
    optional bool is_leader = 3;
    optional tablet.RaftGroupStatePB state = 4;
    optional uint64 num_sst_files = 5;
    optional uint64 num_log_segments = 6;
    optional uint64 num_memtables_intents = 7;
    optional uint64 num_memtables_regular = 8;
  }

  repeated Entry entries = 1;
}

message ImportDataRequestPB {
  optional string tablet_id = 1;
  optional string source_dir = 2;
}

message ImportDataResponsePB {
  // Error message, if any.
  optional TabletServerErrorPB error = 1;
}

message UpdateTransactionRequestPB {
  optional bytes tablet_id = 1;
  optional tablet.TransactionStatePB state = 2;

  optional fixed64 propagated_hybrid_time = 3;
  optional bool is_external = 4;
}

message UpdateTransactionResponsePB {
  // Error message, if any.
  optional TabletServerErrorPB error = 1;

  optional fixed64 propagated_hybrid_time = 2;
}

message GetTransactionStatusRequestPB {
  optional bytes tablet_id = 1;
  repeated bytes transaction_id = 2;
  optional fixed64 propagated_hybrid_time = 3;
  optional fixed64 external_hybrid_time = 4;
}

message GetTransactionStatusResponsePB {
  // Error message, if any.
  optional TabletServerErrorPB error = 1;

  // For backward compatibility we use multiple repeated fields instead of one field
  // with custom message.
  repeated TransactionStatus status = 2;
  // For description of status_hybrid_time see comment in TransactionStatusResult.
  repeated fixed64 status_hybrid_time = 3;

  optional fixed64 propagated_hybrid_time = 4;

  repeated int64 num_replicated_batches = 5;

  // Size could be less than status size. In this case appropriates values should be interpreted
  // as invalid hybrid time.
  repeated fixed64 coordinator_safe_time = 6;

  repeated AbortedSubTransactionSetPB aborted_subtxn_set = 7;
}

message UpdateTransactionWaitingForStatusRequestPB {
  message BlockingTransaction {
    optional bytes transaction_id = 1;
    optional bytes status_tablet_id = 2;
  }
  message WaitingTransaction {
    optional bytes transaction_id = 1;
    repeated BlockingTransaction blocking_transaction = 2;
    optional fixed64 wait_start_time = 3;
  }

  optional fixed64 propagated_hybrid_time = 1;

  optional bytes tablet_id = 2;

  repeated WaitingTransaction waiting_transactions = 3;
}

message UpdateTransactionWaitingForStatusResponsePB {
  optional fixed64 propagated_hybrid_time = 1;

  optional TabletServerErrorPB error = 2;
}

message ProbeTransactionDeadlockRequestPB {
  optional fixed64 propagated_hybrid_time = 1;

  optional bytes tablet_id = 2;

  optional bytes waiting_txn_id = 3;

  optional bytes blocking_txn_id = 4;

  optional bytes detector_uuid = 5;

  optional fixed32 probe_num = 6;

  optional fixed32 min_probe_num = 7;
}

message ProbeTransactionDeadlockResponsePB {
  optional fixed64 propagated_hybrid_time = 1;

  optional TabletServerErrorPB error = 2;

  // If empty, no deadlock was detected. Otherwise, includes reverse ordered list of the deadlock
  // participants detected by the probe recipient. The probe sender should append its own txn to
  // this list before responding to its own probe sender.
  repeated bytes deadlocked_txn_ids = 3;
}

message GetTransactionStatusAtParticipantRequestPB {
  optional bytes tablet_id = 1;
  optional bytes transaction_id = 2;
  optional fixed64 propagated_hybrid_time = 3;
  optional int64 required_num_replicated_batches = 4;
}

message GetTransactionStatusAtParticipantResponsePB {
  // Error message, if any.
  optional TabletServerErrorPB error = 1;
  optional fixed64 propagated_hybrid_time = 2;
  optional int64 num_replicated_batches = 3;
  optional fixed64 status_hybrid_time = 4;
  optional bool aborted = 5;
}

message AbortTransactionRequestPB {
  optional bytes tablet_id = 1;
  optional bytes transaction_id = 2;
  optional fixed64 propagated_hybrid_time = 3;
}

message AbortTransactionResponsePB {
  // Error message, if any.
  optional TabletServerErrorPB error = 1;

  optional TransactionStatus status = 2;
  // For description of status_hybrid_time see comment in TransactionStatusResult.
  optional fixed64 status_hybrid_time = 3;

  optional fixed64 propagated_hybrid_time = 4;

  optional fixed64 coordinator_safe_time = 5;
}

message UpdateTransactionStatusLocationRequestPB {
  optional bytes tablet_id = 1;
  optional bytes transaction_id = 2;
  optional fixed64 propagated_hybrid_time = 3;

  // ID of new status tablet that the transaction will be using.
  optional bytes new_status_tablet_id = 4;
}

message UpdateTransactionStatusLocationResponsePB {
  // Error message, if any.
  optional TabletServerErrorPB error = 1;
}

message GetSplitKeyRequestPB {
  required bytes tablet_id = 1;
  optional fixed64 propagated_hybrid_time = 2;
  optional bool is_manual_split = 3;
}

message GetSplitKeyResponsePB {
  optional TabletServerErrorPB error = 1;
  optional bytes split_partition_key = 2;
  optional bytes split_encoded_key = 3;
  optional fixed64 propagated_hybrid_time = 4;
}

message GetSharedDataRequestPB {
}

message GetSharedDataResponsePB {
  optional bytes data = 1;
}

message GetTserverCatalogVersionInfoRequestPB {
  // We ask for the catalog version of all databases from the tserver.
}

message GetTserverCatalogVersionInfoResponsePB {
  // An entry for each database catalog version info.
  message Entry {
    // The Postgres OID of this database.
    optional uint32 db_oid = 1;
    // The index of the tserver shared memory catalog version array for this database.
    optional int32 shm_index = 2;
    // The current catalog version for this database.
    optional uint64 current_version = 3;
  }

  // Error message, if any.
  optional TabletServerErrorPB error = 1;
  repeated Entry entries = 2;
}
